2022/01/how-to-use-regex-groups-quantifier-in-postgresql.md

```diff
@@ -0,0 +1,86 @@
+tags:
+- postgres
+- sql
+- regex
+- hl7
+github_issue_number: 
+---
+
+I recently encountered a situation where it was necessary to extract address content from a
+specific format (HL7 V2) of text on a PostgreSQL table column. This became possible through applying a regex as the address contains a standard format that regex can match. The following example will be used to demonstrate the process: 
+
+```
+||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|
+```
+### Perl Regex Pattern
+
+In order to manipulate our example, the address section needs to be extracted from the table. The sections and fields are delimited by pipes ( | ) and carets ( ^ ). Uniquely, the address contains eight carets. Therefore, a regex group with alphanumeric and caret repetition is required in order to match our address. The regex is tested through the grep command perl regex option (-P, --perl-regexp) and successfully extracts the address section from the content.
+
+```bash
+$ content='||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|'
+
+$ echo $content | grep -Po '([A-Za-z0-9 ]*\^){8}'
+123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^
+```
+
+### PostgreSQL Regex
+
+PostgreSQL's `regexp_matches` function supports the pattern extraction by matching data from the content. While the same perl regex pattern was added to the `regexp_matches` function, instead of all eight groups, only the eighth value was returned. 
+
+```postgres
+=# SELECT REGEXP_MATCHES('||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|', E'([A-Za-z0-9 ]*\\^){8}[A-Za-z0-9 ]*', 'g');
+ regexp_matches 
+----------------
+ {^}
+(1 row)
+```
+
+More generally, our query is returning the Nth matching group instead of returning all matching groups until the Nth regex group. So if we try to fetch the text matching with 3 groups, the quantifier three will return the third field of all match sections instead of the third group itself.
+
+```postgres
+=# SELECT REGEXP_MATCHES('||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|', E'([A-Za-z0-9 ]*\\^){3}[A-Za-z0-9 ]*', 'g');
+   regexp_matches   
+--------------------
+ {^}
+ {^}
+ {"New York City^"}
+ {USA^}
+(4 rows)
+```
+
+### PLPERL Function
+
+Since the regexp_matches function doesn’t satisfy our requirements, I attempted to use perl regex through the plperl function. The plperl function also created with the perl regex pattern but the plperl regex function resulted in the same behavior.
+
+```postgres
+=# CREATE OR REPLACE FUNCTION perl_regexp_matches (text) RETURNS text AS $$
+    # PL/Perl function body
+    $input = $_[0];
+    ($output) = $input =~ m/([A-Za-z0-9 ]+ ){2}/g;
```

Comment:
The reason this isn't working as you expected is because you're matching multiple times but throwing away all except the first match with the `($output) =`, which is a list context, so the first goes into `$output` and the rest are discarded.

You don't need the `g` global flag since you're just matching a single thing, and you can capture everything you want like this:

```perl
$output = $input =~ m/((?:[A-Za-z0-9 ]+ ){2})/;
```

At least that seems to work right for me. Try it and let me know if I'm missing something.

Reply:
@jonjensen I was able to get the result with following function and updated the post with plperl as one of the solution.
```
=# CREATE OR REPLACE FUNCTION perl_regexp_matches (IN str text, IN pattern text) RETURNS text AS $$
    # PL/Perl function body
    my ($input, $pattern) = @_;
    $output = [$input =~ m/($pattern)/];
    return $output->[0]
$$ LANGUAGE plperl;

```
But I have returned the first value of array since it is production additional value.

```
postgres=# SELECT perl_regexp_matches('||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|', '([A-Za-z0-9 ,/-]*\^){8}[A-Za-z0-9 ]*');
                       perl_regexp_matches
------------------------------------------------------------------
 {"123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York",^}

```